src/main/java/com/iota/iri/service/validation/impl/TransactionSolidifierImpl.java

```diff
@@ -0,0 +1,383 @@
+    private SnapshotProvider snapshotProvider;
+    private TransactionRequester transactionRequester;
+
+    /**
+     * Max size for all queues.
+     */
+    private static final int MAX_SIZE= 10000;
+
+    private static final int SOLIDIFICATION_INTERVAL = 100;
+
+    private static final IntervalLogger log = new IntervalLogger(TransactionSolidifier.class);
+
+    /**
+     * Executor service for running the {@link #processTransactionsToSolidify()}.
+     */
+    private SilentScheduledExecutorService executorService = new DedicatedScheduledExecutorService(
+            "Transaction Solidifier", log.delegate());
+
+    /**
+     * A queue for processing transactions with the {@link #checkSolidity(Hash)} call. Once a transaction has been
+     * marked solid it will be placed into the {@link #transactionsToBroadcast} queue.
+     */
+    private BlockingQueue<Hash> transactionsToSolidify = new ArrayBlockingQueue<>(MAX_SIZE);
+
+    /**
+     * A queue for processing transactions with the {@link #propagateSolidTransactions()} call. This will check
+     * approving transactions with {@link #quickSetSolid(TransactionViewModel)}.
+     */
+    private BlockingQueue<Hash> solidTransactions = new ArrayBlockingQueue<>(MAX_SIZE);
+
+    /**
+     * A set of transactions that will be called by the {@link TransactionProcessingPipeline} to be broadcast to
+     * neighboring nodes.
+     */
+    private BlockingQueue<TransactionViewModel> transactionsToBroadcast = new ArrayBlockingQueue<>(MAX_SIZE);
+
+    private TipsViewModel tipsViewModel;
+
+    /**
+     * Constructor for the solidifier.
+     * @param tangle                    The DB reference
+     * @param snapshotProvider          For fetching entry points for solidity checks
+     * @param transactionRequester      A requester for missing transactions
+     */
+    public TransactionSolidifierImpl(Tangle tangle, SnapshotProvider snapshotProvider, TransactionRequester transactionRequester,
+                                     TipsViewModel tipsViewModel){
+        this.tangle = tangle;
+        this.snapshotProvider = snapshotProvider;
+        this.transactionRequester = transactionRequester;
+        this.tipsViewModel = tipsViewModel;
+    }
+
+    /**
+     *{@inheritDoc}
+     */
+    @Override
+    public void start(){
+        executorService.silentScheduleWithFixedDelay(this::processTransactionsToSolidify, 0,
+                SOLIDIFICATION_INTERVAL, TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     *{@inheritDoc}
+     */
+    @Override
+    public void shutdown() {
+        executorService.shutdownNow();
+    }
+
+    /**
+     *{@inheritDoc}
+     */
+    @Override
+    public void addToSolidificationQueue(Hash hash){
+        try{
+            if(!transactionsToSolidify.contains(hash)) {
+                if (transactionsToSolidify.size() >= MAX_SIZE - 1) {
+                    transactionsToSolidify.remove();
+                }
+
+                transactionsToSolidify.put(hash);
```

Comment:
This was logic that was there for preparation for the milestone epic. I can have it run the way you describe for now and we can re-approach the topic when the epic is ready. The motivation here was to create the separation of powers between state checking milestones, and solidifying transactions. In some instances when `checkSolidity()` is called, it takes a while to run through all the transactions for analysing. This is especially true during larger solidification gaps with a large number of transactions present between milestones. When these calls hang, it slows down the `milestoneSolidifer` because that thread is dependent on the `checkSolidity()` call finishing to continue processing the available transactions in the queue. That includes adding newly seen milestones to the queue. 

I've seen numerous instances during longer syncs, where there will be no message from the solidifier for an extended period of time. When debugging into the nodes to see why that is, I found that the checkSolidity() call was taking a long time to process through these walks and would finish after processing nearly the maximum number of transactions (50000) while doing so. Sometimes these calls could take anywhere between 5 and 20 seconds to process through. Since the queue for the milestone solidifier holds 20 transactions to be put through this call sequentially, there can be long gaps between the solidifier processing new milestones that may have been received through the requests and gossip. The difference is nearly neglible when doing smaller syncs, but over longer periods of time (say a few days for example) it can start to become more noticeable in the later parts of the synchronisation.   

So the logic here was to make sure that the `milestoneSolidifier` is focused on seeing and processing the milestones with minimal interruption, while the `transactionSolidifier` would asynchronously be managing the actual solidification process.

Reply:
I'd put this in a separate PR then
This is not too many lines of code, but still a big important change